occurrence-clustering/src/main/java/org/gbif/occurrence/clustering/RowOccurrenceFeatures.java

```diff
@@ -0,0 +1,200 @@
+package org.gbif.occurrence.clustering;
+
+import com.fasterxml.jackson.databind.JsonNode;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.fasterxml.jackson.module.scala.DefaultScalaModule;
+import org.apache.spark.sql.Row;
+
+import java.io.IOException;
+import java.util.*;
+import java.util.stream.Collectors;
+
+/**
+ * A wrapper around a row giving exposing the terms necessary for clustering.
+ * This allows terms to have a prefix, allowing a Row containing multiple records to be used (e.g. as a result of a
+ * SQL select a.*,b.* from a join b).
+ */
+public class RowOccurrenceFeatures implements OccurrenceFeatures {
+  private static final ObjectMapper OBJECT_MAPPER = new ObjectMapper();
+  static {
+    // required for e.g. correct empty array serializations
+    OBJECT_MAPPER.registerModule(new DefaultScalaModule());
+  }
+
+  private final Row row;
+  private final String prefix; // if field names carry prefixes e.g. t1_scientificName has t1_
+
+  // this hack is only used to correctly serialize fields containing JSON correctly
+  private final Set<String> jsonFields;
+
+  /**
+   * @param row The underlying DataSet row to expose
+   * @param prefix The prefix to strip from field names
+   * @param jsonFields Fields in Row already stored as JSON strings supplied with prefixes.
+   */
+  public RowOccurrenceFeatures(Row row, String prefix, String... jsonFields) {
+    this.row = row;
+    this.prefix = prefix == null ? "" : prefix;
+    this.jsonFields = jsonFields == null ? new HashSet() : new HashSet(Arrays.asList(jsonFields));
+  }
+
+  public RowOccurrenceFeatures(Row row, String prefix) {
+    this(row, prefix, null);
+  }
+
+  public RowOccurrenceFeatures(Row row) {
+    this(row, "");
+  }
+
+  <T> T get(String field) {
+    int fieldIndex = assertNotNull(row.fieldIndex(prefix + field), "Field not found in row schema: "
+      + prefix + field);
```

Comment:
`row.fieldIndex` return an int (not null), and throws an exception if the field doesn't exist.